<a href="https://colab.research.google.com/github/TristanDelgado/Intro-To-Machine-Learning/blob/main/Homework%20Solutions/Homework_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Machine Learning CS-4410-XTIA Homework 7

# Problem - 16.1 (IMAGE RECOGNITION: THE FASHION-MNIST DATASET)**bold text**

## Instructions

 Keras comes bundled with the Fashion-MNIST database of fashion articles which, like the MNIST digits dataset, provides 28-by-28 grayscale images. Fashion-MNIST contains clothing-article images labeled in 10 categories—0 (T-shirt/top), 1 (Trouser), 2 (Pullover), 3 (Dress), 4 (Coat), 5 (Sandal), 6 (Shirt), 7 (Sneaker), 8 (Bag), 9 (Ankle boot)—with 60,000 training samples and 10,000 testing samples. Modify this chapter’s convnet example to load and process Fashion-MNIST rather than MNIST—this requires simply importing the correct module, loading the data then running the model with these images and labels, then re-run the entire example. How well does the model perform on Fashion-MNIST compared to MNIST? How do the training times compare?

10 Categories:

0 (T-shirt/top)

1 (Trouser)

2 (Pullover)

3 (Dress)

4 (Coat)

5 (Sandal)

6 (Shirt)

7 (Sneaker)

8 (Bag)

9 (Ankle boot)


## Solutions


In [1]:
# Importing necessary libraries
from keras.datasets import fashion_mnist
from keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D

In [2]:
# Loading our data
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()
print(f"X_train: {X_train.shape}")
print(f"y_train: {y_train.shape}")
print(f"X_test: {X_test.shape}")
print(f"y_test: {y_test.shape}")

X_train: (60000, 28, 28)
y_train: (60000,)
X_test: (10000, 28, 28)
y_test: (10000,)


In [3]:
# The following code is getting the data into an acceptable form
X_train = X_train.reshape((60000, 28, 28, 1))
print(f"X_train: {X_train.shape}")

X_test = X_test.reshape((10000, 28, 28, 1))
# print(f"X_test: {X_test}")

X_train: (60000, 28, 28, 1)


In [4]:
# Next we will use one-hot encoding to categorize each of the numbers into
# a format our neural network will be able to understand

y_train = to_categorical(y_train)
print(f"y_train shape: {y_train.shape}")
print(f"y_train[0]: {y_train[0]}")

y_test = to_categorical(y_test)

y_train shape: (60000, 10)
y_train[0]: [0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]


In [5]:
# Now we create the neural network

cnn = Sequential()

In [6]:
# Convolution Layer which uses the relationships between pixels that are close to
# one another to learn useful features (or patterns) in small areas of each sample.
# These features become inputs to subsequent layers.

cnn.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu',
                     input_shape=(28, 28, 1)))

/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [7]:
# Next we'll add a "pooling layer" which takes a set number of pixles, finds the largest
# feature and keeps only the largest feature

cnn.add(MaxPooling2D(pool_size=(2, 2)))

In [8]:
# Adding another layer
cnn.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

In [9]:
cnn.add(Flatten())

In [10]:
# Adds the "neurons" of a neural network
cnn.add(Dense(units=128, activation='relu'))

In [11]:
# Adding the final layer
# This is the decision layer, it converts the values into classification values
cnn.add(Dense(units=10, activation='softmax'))

In [12]:
cnn.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                      │ (None, 26, 26, 64)          │             640 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d (MaxPooling2D)         │ (None, 13, 13, 64)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_1 (Conv2D)                    │ (None, 11, 11, 128)         │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_1 (MaxPooling2D)       │ (None, 5, 5, 128)           │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 3200)                │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 128)                 │         409,728 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 10)                  │           1,290 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 485,514 (1.85 MB)

 Trainable params: 485,514 (1.85 MB)

 Non-trainable params: 0 (0.00 B)

In [13]:
# Configure the model for training
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [14]:
# Fitting the model
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 11s 7ms/step - accuracy: 0.7536 - loss: 2.3705 - val_accuracy: 0.8713 - val_loss: 0.3463
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.8777 - loss: 0.3338 - val_accuracy: 0.8883 - val_loss: 0.3042
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.8977 - loss: 0.2762 - val_accuracy: 0.8882 - val_loss: 0.3023
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.9077 - loss: 0.2458 - val_accuracy: 0.8892 - val_loss: 0.3070
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.9129 - loss: 0.2287 - val_accuracy: 0.8927 - val_loss: 0.3038


In [15]:
loss, accuracy = cnn.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8882 - loss: 0.3434


## Question: Answer how well the model performs on Fashion-MNIST compared to MNIST. How do the training times compare?

## Answer: According to the Mnist data which I ran on my own Google Colab while following along with the chapter here are the metrics for MNIST:

MNIST training data:
```
Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 15s 11ms/step - accuracy: 0.8639 - loss: 3.2304 - val_accuracy: 0.9823 - val_loss: 0.0617
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 12s 5ms/step - accuracy: 0.9824 - loss: 0.0526 - val_accuracy: 0.9862 - val_loss: 0.0490
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.9881 - loss: 0.0368 - val_accuracy: 0.9865 - val_loss: 0.0491
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.9904 - loss: 0.0296 - val_accuracy: 0.9848 - val_loss: 0.0541
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.9908 - loss: 0.0285 - val_accuracy: 0.9837 - val_loss: 0.0594
```

Fashion-MNIST training data:
```
Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 11s 7ms/step - accuracy: 0.7536 - loss: 2.3705 - val_accuracy: 0.8713 - val_loss: 0.3463
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.8777 - loss: 0.3338 - val_accuracy: 0.8883 - val_loss: 0.3042
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.8977 - loss: 0.2762 - val_accuracy: 0.8882 - val_loss: 0.3023
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.9077 - loss: 0.2458 - val_accuracy: 0.8892 - val_loss: 0.3070
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.9129 - loss: 0.2287 - val_accuracy: 0.8927 - val_loss: 0.3038
```

As can be seen the training time for the Fashion-MNIST data is a bit faster than the MNIST data. However, the accuracy of the model is significantly less than the MNIST model. At about a 10% differene in accuracy the Fashion-MNIST model would need more training than the MNIST model.

# Problem - 16.4 (CONVNET LAYERS)



## Instructions

Remove the first Dense layer in this chapter’s convnet model. How does this change the prediction accuracy? Several Keras pretrained convnets contain Dense layers with 4096 neurons. Add such a layer before the two Dense layers in this chapter’s convnet model. How does this change the prediction accuracy?

## Solutions

In [18]:
# Importing necessary libraries
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D

### Covnet model with only one dense layer

In [19]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)
cnn = Sequential()
cnn.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Flatten())
cnn.add(Dense(units=10, activation='softmax'))
cnn.summary()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d_2 (Conv2D)                    │ (None, 26, 26, 64)          │             640 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_2 (MaxPooling2D)       │ (None, 13, 13, 64)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_3 (MaxPooling2D)       │ (None, 6, 6, 64)            │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten_1 (Flatten)                  │ (None, 2304)                │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 10)                  │          23,050 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 23,690 (92.54 KB)

 Trainable params: 23,690 (92.54 KB)

 Non-trainable params: 0 (0.00 B)

In [20]:
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)
loss, accuracy = cnn.evaluate(X_test, y_test)

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 6s 5ms/step - accuracy: 0.8332 - loss: 2.8395 - val_accuracy: 0.9707 - val_loss: 0.1791
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.9656 - loss: 0.1802 - val_accuracy: 0.9790 - val_loss: 0.0967
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9755 - loss: 0.1008 - val_accuracy: 0.9720 - val_loss: 0.1074
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9780 - loss: 0.0787 - val_accuracy: 0.9798 - val_loss: 0.0859
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9796 - loss: 0.0706 - val_accuracy: 0.9768 - val_loss: 0.0898
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9672 - loss: 0.1361


Initially the accuracy of the covnet model was 0.9794 unchanged. After the removal of the Dense layer with 128 neurons the accuracy decreased by a small degree to 0.9672. Not a big dip surprisingly.

### Covnet Model with an extra Dense layer with 4096 neurons

In [21]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)
cnn = Sequential()
cnn.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Flatten())
cnn.add(Dense(units=4096, activation='relu'))
cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))
cnn.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d_3 (Conv2D)                    │ (None, 26, 26, 64)          │             640 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_4 (MaxPooling2D)       │ (None, 13, 13, 64)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_5 (MaxPooling2D)       │ (None, 6, 6, 64)            │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten_2 (Flatten)                  │ (None, 2304)                │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 4096)                │       9,441,280 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_4 (Dense)                      │ (None, 128)                 │         524,416 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_5 (Dense)                      │ (None, 10)                  │           1,290 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 9,967,626 (38.02 MB)

 Trainable params: 9,967,626 (38.02 MB)

 Non-trainable params: 0 (0.00 B)

In [22]:
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)
loss, accuracy = cnn.evaluate(X_test, y_test)

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 10s 8ms/step - accuracy: 0.8109 - loss: 9.4874 - val_accuracy: 0.9702 - val_loss: 0.1158
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 7s 6ms/step - accuracy: 0.9684 - loss: 0.1173 - val_accuracy: 0.9778 - val_loss: 0.0900
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.9836 - loss: 0.0543 - val_accuracy: 0.9810 - val_loss: 0.0690
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.9878 - loss: 0.0425 - val_accuracy: 0.9842 - val_loss: 0.0547
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.9897 - loss: 0.0328 - val_accuracy: 0.9823 - val_loss: 0.0648
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9751 - loss: 0.0953


Initially the accuracy of the covnet model was 0.9794 unchanged. After adding another Dense layer with 4096 new neurons the accuracty decreased although by a very small machine. The new accuracy was 0.9751. A 0.0043 percent difference in accuracy.